### ISSUE TITLE
Git LFS asks for authentication after every lfs-tracked file

### ISSUE THREAD

**Issue by AndreTInfante** [github/git-lfs#858]
My team is working on a project using Github and LFS, and having a fairly serious problem. One of our team members is having an issue with LFS where, when he clones the repo, he gets asked to authenticate after every lfs-tracked file. As we have several hundred lfs-tracked files, this has rendered Git pretty much unuseable.

The issue seems to come up while using both https and ssh. 

Any ideas what might be the problem?
Thank you,
Andre

**Comment by technoweenie** [comment]
Set up a credential cacher: https://help.github.com/articles/caching-your-github-password-in-git/.

Though, you shouldn't need it with SSH stuff, since that should be generating short-lived authentication tokens.

**Comment by AndreTInfante** [comment]
"git config --global credential.helper wincred" exits silently, and doesn't seem to help. Is there something extra needed to make that work?

He's on windows 10, using the powershell git shell if that matters.

**Comment by SindreOpsahl** [comment]
I'm the team member facing this problem. 
I've already set up SSH following this: https://help.github.com/articles/generating-ssh-keys/
And I followed the instructions in the link you provided already.

What is happening is that whenever I do a clone or a pull is that it seems to succesfully clone the project, but when it ostensibly gets to the stage where it downloads the lfs assets it needs to authenticate for each of them. 

This is what my GitBash looks like. It runs fine until the last two lines which each popped up the OpenSSH window seen on the side there. This is with cloning the ssh url. Using the HTTPS url the prompt asks for username and password. 

![git-fail](https://cloud.githubusercontent.com/assets/15728386/11355415/33a0edc4-9255-11e5-9061-d6151abd7744.png)

When I try to clone with GitHub Desktop and SourceTree a similar thing happens where it succesfully clones up to a certain stage and then aborts. I assume because of the authentication. 
Sourcetree returns this error message:

```
Error accessing media: Assets/PBS Materials Variety Pack/Carpet/Textures/Carpet 1 bump.png (52efb6e74025f33515d042216780a85734e108b23c89f147728796d371e44d30)

Errors logged to .git\lfs\objects\logs\20151123T184439.5305253.log
Use `git lfs logs last` to view the log.
error: external filter git-lfs smudge %f failed 2
error: external filter git-lfs smudge %f failed
fatal: Assets/PBS Materials Variety Pack/Carpet/Textures/Carpet 1 bump.png: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

**Comment by technoweenie** [comment]
You can setup an ssh agent to temporarily cache the ssh passphrase prompts. I don't use SSH often other than for testing (and definitely not on Windows), so I won't be much help there unfortunately. Another option is to use an SSH key without a passphrase, at the expense of security.

As for SourceTree, check out ".git\lfs\objects\logs\20151123T184439.5305253.log" for the stacktrace that git lfs saved. #846 and #853 are existing open issues regarding SourceTree and system path issues. Unfortunately, these are issues that only SourceTree can solve AFAIK. 

If the HTTPS credential cacher isn't working, make sure your git version is above 1.8.2. Git LFS uses the `git-credential` tool to get credentials for the API calls. Also, it only affects HTTPS remotes, not SSH remotes.

Here's how you can test if the git credential helper is working:

```bash
# in Git Bash:

# if this prompts for username, nothing is cached
$ echo "protocol=https\nhost=github.com" | git credential fill

# this clears any cached settings for the host
$ echo "protocol=https\nhost=github.com" | git credential fill | git credential reject

# this attempts to save to your credential helper:
$ echo "protocol=https\nhost=github.com" | git credential fill | git credential approve
```

This is essentially all Git LFS is doing ^

**Comment by SindreOpsahl** [comment]
Hm ok, `$ echo "protocol=https\nhost=github.com" | git credential fill` does prompt for a username so i guess it didn't get cached.

**Comment by SindreOpsahl** [comment]
Argh, clearing and trying to save using the commands you used don't seem to work. I'm at the head revision with missing files, so I'm using `checkout -f HEAD`, but like you said it only works with HTTPS so I suppose I will have to do a new clone.

**Comment by technoweenie** [comment]
No need to re-clone it. Try:

```bash
$ git remote rm origin
$ git remote add origin https://github.com/....
```

**Comment by SindreOpsahl** [comment]
Oh I already started the process. Luckily my internet is reasonably fast so it doesn't too long. 

By the way, were the reject and approve commands above supposed to give any kind of feedback? I was prompted to give my username and password but other than that not much happened. I'm new to git and making sure I do things right.

**Comment by technoweenie** [comment]
Nope, those are git "plumbing" commands, designed to be called by other git commands. You shouldn't have to approve or reject anything manually though. Git should be doing it for you automatically. In fact, if it's setup properly, Git LFS won't ever prompt you because Git should've already asked, and approved it when you typed `git clone/push/pull`.

**Comment by SindreOpsahl** [comment]
Yeah that was my impression too, that lfs just silently runs in the background. Which is why these very aggressive errors are puzzling to me. 

But on the bright side cloning from HTTPS seems to have been the key! The files are now downloading without interruption! Thanks a lot for your help.

Do you have any idea what could have caused this issue in the first place? So we (and anyone reading this) can do the right steps during setup to avoid similar problems in the future.

**Comment by technoweenie** [comment]
Great, glad to hear you got this working. 

The core issues you experienced are core Git things, not necessarily related to LFS: SSH, HTTP credentials, native Git tools. I'm seeing a lot of progress for Git on Windows though, so hopefully the story is very different a year from now.

* LFS is a young project. Definitely needs real documentation outside of the command man pages.
* Atlassian contributes to the Git LFS project, so I imagine Sourcetree's LFS support will improve.
* MS is doing a lot with Git and LFS, including a [new credential manager](https://github.com/Microsoft/Git-Credential-Manager-for-Windows), and even [openssh support](http://www.hanselman.com/blog/ItsHappeningOpenSSHForWindowsfromMicrosoft.aspx).

**Comment by guillaume-chevalier** [comment]
Seems related to: 
https://github.com/git-lfs/git-lfs/issues/203

**Comment by viperscape** [comment]
I am having this problem, using ssh with a key (on Windows), I am prompted during push/pull as usual in bash, but then again with a popup for each lfs file in question. Are not many people using keyed ssh and git-lfs?
Using Ubuntu on Windows, git lfs asks for login on each file with https as the remote; so it seems like git-lfs does not operate similarly to git when fetching files, regardless of OS or protocol. I wonder why there was a deviation from how it operates compared to git itself.

**Comment by viperscape** [comment]
As a workaround, in case someone else wants to resolve this in the mean time, I used Putty's Pageant and PLINK tools, which will basically push the key for you. So far so good. [Here is some information on it](https://stackoverflow.com/a/43313491) and [here](https://stackoverflow.com/a/35116084)

**Comment by Jamesernator** [comment]
This is the biggest pain point for me with `git-lfs`, every single time I edit files the workflow tends to be:

1. Do git pull ssh asks for password
2. Enter password
3. Git LFS begins downloading then on the first file "Damn it I forgot to use ssh-add"
4. Kill git but git repository is now in a broken state because git lfs didn't finish
5. Run ssh-add then git pull again, discovering which files are broken
6. Delete the broken files and finally git pull again

Frankly this gets quite tedious and I'm rather sick of as I forget to use `ssh-add` **every single time**.

**Comment by drew-gross** [comment]
I also have this problem. For me it's more like:

1. Attempt to checkout a branch that I have never checked out before (remote only)
2. Realize there are git lfs files and I haven't done ssh-add yet
3. Kill the command so I can do ssh-add
4. Repository is in broken state

90% of the time I can checkout just fine, because there are no changes to lfs files, but when I can't, it's really annoying.

**Comment by technoweenie** [comment]
You should be able to run `git checkout .` or `git reset --hard` (`reset` will revert any uncommitted changes though) to fix the working directory.

As for SSH: LFS does cache results from the `git-lfs-authenticate` command, though that's a relatively newer feature in LFS 2.3 I think. Is your ssh service returning an expiration?

If your remote is `git@github.com:foo/bar.git`, you can run this ssh command:

```sh
$ ssh git@github.com git-lfs-authenticate foo/bar.git download
{
  "href": "https://lfs.github.com/foo/bar",
  "header": {
    "Authorization": "RemoteAuth something"
  },
  "expires_at": "2017-10-03T22:45:39Z",
  "expires_in": 599
}
```

Does that response include `expires_at` or `expires_in`?

**Comment by drew-gross** [comment]
Yes both of those are set

**Comment by technoweenie** [comment]
Are you using LFS 2.3.0+? Is [`lfs.cachecredentials`](https://github.com/git-lfs/git-lfs/blob/986f8a25dbc272545158744cbd869fb0b519dcaa/docs/man/git-lfs-config.5.ronn#L60-L63) disabled?

**Comment by instantname** [comment]
Caching credentials is good, but it may be more powerful to do everything in a single session. For example, what if access is authorized by one-time passwords?

**Comment by ttaylorr** [comment]
> Caching credentials is good, but it may be more powerful to do everything in a single session. For example, what if access is authorized by one-time passwords?

Hm, can you clarify what you mean by one-time passwords? My understanding is that you're talking about a credential issued that is valid for the duration of the LFS session, in which case `lfs.cachecredentials` would be a good fit.

**Comment by instantname** [comment]
I was thinking of [libpam-yubico](https://developers.yubico.com/yubico-pam/), a PAM module that you can include for example in the SSH authentication process. The password has to be different each time you gave a successful one to the server. I don't know how ``lfs-cachecredentials`` works, but if it does not store passwords themselves, I guess that should be OK for OTPs.

**Comment by ttaylorr** [comment]
@instantname I think that would be something that an implementor of the LFS API specification would have to handle, not the client itself, no?
